src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +185,64 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            List<Socket> sockets = new List<Socket>(serverAddresses.Length);
+            List<Task> connectTasks = new List<Task>(serverAddresses.Length);
+
+            ExceptionDispatchInfo lastSocketConnectEx = null;
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port);
+
+                    sockets.Add(socket);
+                    connectTasks.Add(socketTask);
                 }
                 catch (Exception exc)
                 {
                     socket.Dispose();
-                    lastException = ExceptionDispatchInfo.Capture(exc);
+                    lastSocketConnectEx = ExceptionDispatchInfo.Capture(exc);
                 }
             }
 
-            // Propagate the last failure that occurred
-            if (lastException != null)
+            if (sockets.Count == 0)
             {
-                lastException.Throw();
+                Debug.Assert(lastSocketConnectEx != null, "Expected to see an exception if we encountered invalid socket count.");
+                lastSocketConnectEx.Throw();
             }
 
-            // Should never get here.  Either there will have been no addresses and we'll have thrown
-            // at the beginning, or one of the addresses will have worked and we'll have returned, or
-            // at least one of the addresses will failed, in which case we will have propagated that.
-            throw new ArgumentException();
+            // Wait on all connection tasks, and return socket associated with first completed task.
+            return Task.Run(
+                () => {
+                    Socket resultSocket = null;
+                    ExceptionDispatchInfo connectWaitEx = null;
+                    try
+                    {
+                        int taskIndex = Task.WaitAny(connectTasks.ToArray());
+                        resultSocket = sockets[taskIndex];
+                        sockets.RemoveAt(taskIndex);
+                    }
+                    catch (Exception exc)
+                    {
+                        connectWaitEx = ExceptionDispatchInfo.Capture(exc);
+                    }
+                    finally
+                    {
+                        foreach (Socket socket in sockets)
+                        {
+                            socket.Dispose();
```

Comment:
If we have a connected socket, then the other sockets aren't necessary. Is there an explicit call to cancel the async connections for these unneeded sockets? Just socket.Shutdown()?